store/pg_engine.go

```diff
@@ -227,91 +217,146 @@ func (db *DB) seedFile(name string) error {
+	})
 
-	maxVer := versionFromInt(db.schemaVersion)
-	for _, name := range names {
-		versionPrefix := strings.Split(filepath.Base(name), "__")[0]
-		version, err := strconv.Atoi(versionPrefix)
+	for _, version := range versions {
+		// If the migration version is greather than software schema version, we will skip the migration.
+		if version.GT(db.schemaVersion) {
+			db.l.Info(fmt.Sprintf("Skip this migration: %s; the corresponding migration version %s is bigger than maximum schema version %s.", version, version, db.schemaVersion))
+			continue
+		}
+		// If the migration version is less than or equal to the current version, we will skip the migration since it's already applied.
+		if version.LE(*curVer) {
+			db.l.Info(fmt.Sprintf("Skip this migration: %s; the current migration version is %s.", version, *curVer))
+			continue
+		}
+
+		// Migrate migration files.
+		db.l.Info(fmt.Sprintf("Migrating %s...", version))
+		names, err := fs.Glob(migrationFS, fmt.Sprintf("migration/%s/*.sql", version))
 		if err != nil {
-			return fmt.Errorf("invalid migration file format %s, expected number prefix", filepath.Base(name))
+			return err
 		}
-		v := versionFromInt(version)
-		if v.biggerThan(maxVer) {
-			db.l.Debug(fmt.Sprintf("Skip this migration file: %s. The corresponding migration version %s is bigger than maximum schema version %s.", name, v, maxVer))
-		} else if v.biggerThan(curVer) {
-			// Migrate migration files.
-			db.l.Info(fmt.Sprintf("Migrating %s...", name))
+		var stmtBuf bytes.Buffer
+		var baseNames []string
+		for _, name := range names {
+			// Skip the latest sql file.
+			baseName := filepath.Base(name)
+			if baseName == latestSchemaFile {
+				continue
+			}
+			if baseName == latestDataFile {
+				continue
+			}
+			baseNames = append(baseNames, baseName)
 			buf, err := fs.ReadFile(migrationFS, name)
 			if err != nil {
-				return err
+				return fmt.Errorf("failed to read migration file %q, error %w", name, err)
 			}
-			// TODO(d): support semantic versioning.
-			if _, _, err := d.ExecuteMigration(
-				ctx,
-				&dbdriver.MigrationInfo{
-					ReleaseVersion: db.serverVersion,
-					Version:        fmt.Sprintf("%d", version),
-					Namespace:      databaseName,
-					Database:       databaseName,
-					Environment:    "", /* unused in execute migration */
-					Source:         dbdriver.LIBRARY,
-					Type:           dbdriver.Migrate,
-					Description:    fmt.Sprintf("Migrate %s.", filepath.Base(name)),
-				},
-				string(buf),
-			); err != nil {
-				return fmt.Errorf("failed to migrate schema version %q, error: %v", filepath.Base(name), err)
+			if _, err := stmtBuf.Write(buf); err != nil {
+				return fmt.Errorf("failed to write buffer for migration file %q, error %w", name, err)
 			}
-		} else {
-			db.l.Info(fmt.Sprintf("Skip this migration file: %s. The corresponding migration version %s has already been applied.", name, v))
+			if _, err := stmtBuf.WriteString("\n\n"); err != nil {
```

Comment:
This brings me the idea of leveraging the PATCH version actually. e.g. We can assign patch number to the migration files under 1.1.0.

Thus we don't need to combine the migration files and do them together. Each migration file should be self contained.

And migration files under a major version might also have dependency, so they need to design a scheme like prefix with number anyway.

wdyt?

Reply:
Yeah, I thought about the option, however, we do like to align software releases together with schema releases. For example, software 1.0.0 should include the schema+data for 1.0.0 instead of 1.0.1.

Reply:
I would view whether to couple software release with schema release is up to application owners. While for us, we could enforce a workflow to let the schema version drives the software version.

So if schema bumps to 1.0.1, then the software version must be 1.0.1. Not the other way around.